tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -4769,61 +4775,84 @@ Status ConvertPad(OpConverterParams* params) {
   }
 
-  // Only supports padding on less than 2 axis GIE-2579
-  if (pad_index.size() > 2) {
+  // Only supports padding on less equal than 2 axis GIE-2579
+  // TODO(tfeher) This could be improved by adding multiply TRT pad layers
+  if (tf_pad_index.size() > 2) {
     return errors::InvalidArgument(
         "Padding layer does not support padding on > 2");
   }
 
   // Padding on batch dimension is not supported
-  if (pad_index[0] == 0) {
+  if (params->use_implicit_batch && tf_pad_index[0] == 0) {
     return errors::InvalidArgument(
         "Padding layer does not support padding on batch dimension");
   }
 
-  // Not doing the legit thing here. ignoring padding on dim 1 and 3;
-  // TODO(jie): implement pad as uff parser
-  if (pad_index.size() == 2 && pad_index[0] == 0 && pad_index[1] == 3) {
-    return errors::Unimplemented(
-        "Padding layer does not support padding on dimension 1 and 3 yet");
-  }
   if (params->validation_only) return Status::OK();
 
-  bool legit_pad = true;
+  // TRT can only do the padding at the last two dimensions. We transpose the
+  // input tensor if needed.
+  bool transposed_pad = false;
+  std::vector<int> transpose_idx(nb_dims);
+  std::iota(transpose_idx.begin(), transpose_idx.end(), 0);
+
+  // trt_pad_index denotes the actual idx where the padding is performed by TRT.
+  std::vector<int> trt_pad_index{nb_dims - 2, nb_dims - 1};
+  bool trt_pad_index_transposed = false;
+
+  // How many zeros are padded at the last two dimensions.
   nvinfer1::DimsHW pre_padding(0, 0);
   nvinfer1::DimsHW post_padding(0, 0);
 
-  std::vector<int32_t> permuted_pad_index(pad_index);
-  if (pad_index[0] == 1) {
-    legit_pad = false;
-    TF_RETURN_IF_ERROR(params->converter->TransposeTensor(
-        tensor, {0, 3, 2, 1}, &tensor, node_def, "to_pad"));
-    permuted_pad_index[0] = 3;
-  }
-
-  for (size_t i = 0; i < pad_index.size(); i++) {
-    int index = pad_index[i];
-    if (permuted_pad_index[i] == 2) {
-      pre_padding.h() = pad_data[index * 2];
-      post_padding.h() = pad_data[index * 2 + 1];
-    } else if (permuted_pad_index[i] == 3) {
-      pre_padding.w() = pad_data[index * 2];
-      post_padding.w() = pad_data[index * 2 + 1];
+  // Two special cases where we can avoid permutations.
+  if (tf_pad_index.size() == 1 && tf_pad_index[0] == nb_dims - 1) {
+    // In case there is only one dim to pad, and tf_index is the last dim,
+    // then we set trt_idx to be the same dim.
+    trt_pad_index[0] = nb_dims - 1;
+  }
+  if (tf_pad_index.size() == 2 && tf_pad_index[1] == nb_dims - 2) {
+    // Assume an NCHW tensor, where padding is needed at CH, i.e. at dims [1,2].
+    // The trt_pad_index array contains values [2, 3]. Here we transpose
+    // trt_pad_index to be [3, 2]. This way the last dims are the same
+    // (trt_pad_index[1] == tf_pad_index[1]). The aim is to reduce the number of
+    // permutations: now only one transpose will be needed from trt_pad_index[0
+    // to tf_pad_index[0].
```

Comment:
Let's revise the comment:
tf_pad_index only has two values that are in ascending order. If tf_pad_index[1] is nb_dims-2, then swapping the two values in trt_pad_index here makes it possible to only swapping one pair of dimensions (tf_pad_index[0], nb_dims-1) in the input tensor. Otherwise, we would have to swap two pairs of dimensions in the input tensor: (tf_pad_index[0], nb_dims-2), (tf_pad_index[1], nb_dims-1). Here is an example for a 4D input tensor:
tf_pad_index = [1, 2]
trt_pad_index = [3, 2]
transpose_idx = [0, 3, 2, 1]

Reply:
Changed it.